File: frontend/src/types/webinar.ts
/** Webinar & Landing Page types — match backend Pydantic schemas (camelCase wire format). */

export type WebinarStatus = 'draft' | 'scheduled' | 'live' | 'completed' | 'cancelled';
export type MeetingProvider = 'none' | 'zoom' | 'google_meet';
export type LandingPageStatus = 'draft' | 'published' | 'archived';
export type LandingPageType = 'opt_in' | 'thank_you' | 'sales' | 'replay' | 'custom';

export interface Webinar {
  id: string;
  organization_id: string;
  created_by: string;
  title: string;
  slug: string;
  description: string | null;
  status: WebinarStatus;
  provider: MeetingProvider;
  provider_meeting_id: string | null;
  starts_at: string | null;
  ends_at: string | null;
  timezone: string;
  capacity: number | null;
  location_type: string;
  agenda: Record<string, unknown> | null;
  ai_topic: string | null;
  is_published: boolean;
  registration_count: number;
  attendance_count: number;
  visitor_count: number;
  is_paid: boolean;
  price_cents: number;
  currency: string;
  payment_gateway: string;
  created_at: string;
  updated_at: string;
}

export interface WebinarCreatePayload {
  title: string;
  description?: string | null;
  status?: WebinarStatus;
  provider?: MeetingProvider;
  provider_meeting_id?: string | null;
  starts_at?: string | null;
  ends_at?: string | null;
  timezone?: string;
  capacity?: number | null;
  location_type?: string;
  agenda?: Record<string, unknown> | null;
  ai_topic?: string | null;
  is_published?: boolean;
  is_paid?: boolean;
  price_cents?: number;
  currency?: string;
  payment_gateway?: string;
}

export interface WebinarUpdatePayload {
  title?: string;
  description?: string | null;
  status?: WebinarStatus;
  provider?: MeetingProvider;
  provider_meeting_id?: string | null;
  starts_at?: string | null;
  ends_at?: string | null;
  timezone?: string;
  capacity?: number | null;
  location_type?: string;
  agenda?: Record<string, unknown> | null;
  ai_topic?: string | null;
  is_published?: boolean;
  is_paid?: boolean;
  price_cents?: number;
  currency?: string;
  payment_gateway?: string;
}

export interface PaginatedWebinars {
  items: Webinar[];
  total: number;
  limit: number;
  offset: number;
}

export interface WebinarDuplicateResponse {
  original_id: string;
  duplicate_id: string;
  duplicate_slug: string;
}

// --- Landing Page types ---

export interface LandingPage {
  id: string;
  webinar_id: string;
  organization_id: string;
  created_by: string;
  title: string;
  slug: string;
  status: LandingPageStatus;
  page_type: LandingPageType;
  content: Record<string, any> | null;
  meta_title: string | null;
  meta_description: string | null;
  meta_image: string | null;
  custom_head_html: string | null;
  custom_body_html: string | null;
  published_at: string | null;
  is_published: boolean;
  template_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface LandingPageCreatePayload {
  webinar_id: string;
  title: string;
  slug?: string;
  page_type?: LandingPageType;
  content?: Record<string, unknown> | null;
  meta_title?: string | null;
  meta_description?: string | null;
  meta_image?: string | null;
  custom_head_html?: string | null;
  custom_body_html?: string | null;
  is_published?: boolean;
  template_id?: string | null;
}

export interface LandingPageUpdatePayload {
  title?: string;
  slug?: string;
  page_type?: LandingPageType;
  content?: Record<string, unknown> | null;
  meta_title?: string | null;
  meta_description?: string | null;
  meta_image?: string | null;
  custom_head_html?: string | null;
  custom_body_html?: string | null;
  is_published?: boolean;
  template_id?: string | null;
}

export interface LandingPageItem {
  id: string;
  webinar_id: string;
  title: string;
  slug: string;
  status: LandingPageStatus;
  page_type: LandingPageType;
  is_published: boolean;
  template_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface LandingPageDetail extends LandingPageItem {
  webinar_id: string;
  organization_id: string;
  created_by: string;
  content: Record<string, any> | null;
  meta_title: string | null;
  meta_description: string | null;
  meta_image: string | null;
  custom_head_html: string | null;
  custom_body_html: string | null;
  published_at: string | null;
}

export interface PaginatedLandingPages {
  items: LandingPageItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface LandingPageDuplicateResponse {
  original_id: string;
  duplicate_id: string;
  duplicate_slug: string;
}

export interface LandingPageStats {
  id: string;
  title: string;
  slug: string;
  page_type: LandingPageType;
  is_published: boolean;
  visit_count: number;
  registration_count: number;
  conversion_rate: number;
}

File: backend/app/schemas/webinar.py
"""Pydantic schemas for Webinar CRUD.

Wire convention is camelCase (matching the auth schemas); however the Webinar
ORM columns themselves are snake_case so we keep pydantic field names as-is and
rely on attribute access via ``from_attributes=True``. Enum fields serialize to
their string value (``use_enum_values=True``) so the API returns plain strings.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from app.models.webinar import MeetingProvider, RegistrantStatus, WebinarStatus


# --- Create / Update payloads ------------------------------------------------- #


class WebinarCreate(BaseModel):
    """Payload for creating a webinar. Organization is resolved from X-Organization-Id."""

    title: str = Field(..., min_length=1, max_length=255, description="Webinar title")
    description: str | None = Field(default=None, max_length=4000)
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    timezone: str = Field(default="UTC", max_length=64)
    capacity: int | None = Field(default=None, ge=1)
    status: WebinarStatus = Field(default=WebinarStatus.draft)
    provider: MeetingProvider = Field(default=MeetingProvider.none)
    location_type: str = Field(default="online", max_length=16)
    ai_topic: str | None = Field(default=None, max_length=512)
    is_published: bool = False
    is_paid: bool = False
    price_cents: int = Field(default=0, ge=0, description="Price in cents (e.g. 1999 = $19.99). Required > 0 when is_paid=True.")
    currency: str = Field(default="usd", max_length=3, description="ISO 4217 currency code, e.g. usd, inr, eur")
    payment_gateway: str = Field(default="stripe", max_length=16, description="Payment gateway: stripe or razorpay")

    @model_validator(mode="after")
    def _price_required_when_paid(self):
        if self.is_paid and self.price_cents <= 0:
            raise ValueError("price_cents must be greater than 0 for paid webinars")
        return self

    @field_validator("title")
    @classmethod
    def _strip_title(cls, v: str) -> str:
        return v.strip()


class WebinarUpdate(BaseModel):
    """Payload for updating a webinar. All fields optional; org scoping enforced by endpoint."""

    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=4000)
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    timezone: str | None = Field(default=None, max_length=64)
    capacity: int | None = Field(default=None, ge=1)
    status: WebinarStatus | None = None
    provider: MeetingProvider | None = None
    location_type: str | None = Field(default=None, max_length=16)
    ai_topic: str | None = Field(default=None, max_length=512)
    is_published: bool | None = None
    is_paid: bool | None = None
    price_cents: int | None = Field(default=None, ge=0)
    currency: str | None = Field(default=None, max_length=3)
    payment_gateway: str | None = Field(default=None, max_length=16)

    @model_validator(mode="after")
    def _price_required_when_paid(self):
        if self.is_paid is True and self.price_cents is not None and self.price_cents <= 0:
            raise ValueError("price_cents must be greater than 0 for paid webinars")
        return self


# --- Read / List schemas ----------------------------------------------------- #


_READ_CONFIG = ConfigDict(from_attributes=True, use_enum_values=True)


class WebinarItem(BaseModel):
    """Compact list item — used in paginated responses."""

    model_config = _READ_CONFIG

    id: UUID
    organization_id: UUID
    title: str
    slug: str
    status: WebinarStatus
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    capacity: int | None = None
    is_published: bool = False
    is_paid: bool = False
    price_cents: int | None = 0
    currency: str | None = "usd"
    payment_gateway: str | None = "stripe"
    # Denormalized counters — updated on every registration/attendance change
    registration_count: int = 0
    attendance_count: int = 0
    visitor_count: int = 0
    created_at: datetime
    updated_at: datetime


class WebinarDetail(BaseModel):
    """Full webinar detail — includes all fields plus agenda JSON."""

    model_config = _READ_CONFIG

    id: UUID
    organization_id: UUID
    created_by: UUID
    title: str
    slug: str
    description: str | None = None
    status: WebinarStatus
    provider: MeetingProvider | None = MeetingProvider.none
    provider_meeting_id: str | None = None
    starts_at: datetime | None = None
    ends_at: datetime | None = None
    timezone: str = "UTC"
    capacity: int | None = None
    location_type: str = "online"
    agenda: dict[str, Any] | None = None
    ai_topic: str | None = None
    is_published: bool = False
    is_paid: bool = False
    price_cents: int | None = 0
    currency: str | None = "usd"
    payment_gateway: str | None = "stripe"
    registration_count: int = 0
    attendance_count: int = 0
    visitor_count: int = 0
    created_at: datetime
    updated_at: datetime


class WebinarListResponse(BaseModel):
    """Paginated list envelope."""

    items: list[WebinarItem]
    total: int
    limit: int
    offset: int


class WebinarDuplicateResponse(BaseModel):
    """Result of duplicating a webinar."""

    original_id: UUID
    duplicate_id: UUID
    duplicate_slug: str


# --- Registrant schemas ------------------------------------------------------ #


class RegistrantCreate(BaseModel):
    """Public registrant payload."""

    email: str = Field(..., max_length=255, description="Registrant email")
    first_name: str | None = Field(default=None, max_length=128)
    last_name: str | None = Field(default=None, max_length=128)
    custom_fields: dict[str, Any] | None = None
    utm_source: str | None = Field(default=None, max_length=128)
    utm_medium: str | None = Field(default=None, max_length=128)
    utm_campaign: str | None = Field(default=None, max_length=128)
    referrer: str | None = Field(default=None, max_length=512)
    landing_page_id: UUID | None = None

    @field_validator("email")
    @classmethod
    def _normalize_email(cls, v: str) -> str:
        return v.strip().lower()


class RegistrantItem(BaseModel):
    """Registrant detail for admin list."""

    model_config = _READ_CONFIG

    id: UUID
    webinar_id: UUID
    email: str
    first_name: str | None
    last_name: str | None
    status: RegistrantStatus
    registered_at: datetime
    attended: bool
    attended_at: datetime | None
    join_token: str
    custom_fields: dict[str, Any] | None


class RegistrantListResponse(BaseModel):
    """Paginated registrants list."""

    items: list[RegistrantItem]
    total: int
    limit: int
    offset: int


class RegistrationResponse(BaseModel):
    """Response after successful public registration."""

    id: UUID
    webinar_id: UUID
    email: str
    join_url: str
    status: RegistrantStatus
    registered_at: datetime


# --- Attendance schemas ------------------------------------------------------ #


class AttendanceCheckIn(BaseModel):
    """Payload to record attendee presence."""

    join_token: str = Field(..., min_length=1)
    duration_seconds: int = Field(default=0, ge=0)


class AttendanceItem(BaseModel):
    """Attendance record."""

    model_config = _READ_CONFIG

    id: UUID
    webinar_id: UUID
    registrant_id: UUID
    joined_at: datetime
    left_at: datetime | None
    duration_seconds: int